tests/diff/binary.c

```diff
@@ -261,3 +261,232 @@ void test_diff_binary__delta_append(void)
+	git_diff_free(diff);
+	git_index_free(index);
+}
+
+static int print_cb(
+	const git_diff_delta *delta,
+	const git_diff_hunk *hunk,
+	const git_diff_line *line,
+	void *payload)
+{
+	git_buf *buf = (git_buf *)payload;
+
+	if (hunk)
+		git_buf_put(buf, hunk->header, hunk->header_len);
+
+	if (line)
+		git_buf_put(buf, line->content, line->content_len);
+
+	return git_buf_oom(buf) ? -1 : 0;
+}
+
+void test_diff_binary__print_patch_from_diff(void)
+{
+	git_index *index;
+	git_diff *diff;
+	git_buf actual = GIT_BUF_INIT;
+	git_diff_options opts = GIT_DIFF_OPTIONS_INIT;
+	const char *expected =
+		"diff --git a/untimely.txt b/untimely.txt\n" \
+		"index 9a69d960ae94b060f56c2a8702545e2bb1abb935..1111d4f11f4b35bf6759e0fb714fe09731ef0840 100644\n" \
+		"GIT binary patch\n" \
+		"delta 32\n" \
+		"nc%1vf+QYWt3zLL@hC)e3Vu?a>QDRl4f_G*?PG(-ZA}<#J$+QbW\n" \
+		"\n" \
+		"delta 7\n" \
+		"Oc%18D`@*{63ljhg(E~C7\n";
+
+	opts.flags = GIT_DIFF_SHOW_BINARY | GIT_DIFF_FORCE_BINARY;
+	opts.id_abbrev = GIT_OID_HEXSZ;
+
+	repo = cl_git_sandbox_init("renames");
+	cl_git_pass(git_repository_index(&index, repo));
+
+	cl_git_append2file("renames/untimely.txt", "Oh that crazy Kipling!\r\n");
+
+	cl_git_pass(git_diff_index_to_workdir(&diff, repo, index, &opts));
+
+	cl_git_pass(git_diff_print(diff, GIT_DIFF_FORMAT_PATCH, print_cb, &actual));
+
+	cl_assert_equal_s(expected, actual.ptr);
+
+	git_buf_free(&actual);
+	git_diff_free(diff);
+	git_index_free(index);
+}
+
+struct diff_data {
+	char *old_path;
+	git_oid old_id;
+	git_buf old_binary_base85;
+	size_t old_binary_inflatedlen;
+	git_diff_binary_t old_binary_type;
+
+	char *new_path;
+	git_oid new_id;
+	git_buf new_binary_base85;
+	size_t new_binary_inflatedlen;
+	git_diff_binary_t new_binary_type;
+};
+
+static int file_cb(
+	const git_diff_delta *delta,
+	float progress,
+	void *payload)
+{
+	struct diff_data *diff_data = payload;
+
+	if (delta->old_file.path)
+		diff_data->old_path = git__strdup(delta->old_file.path);
+
+	if (delta->new_file.path)
+		diff_data->new_path = git__strdup(delta->new_file.path);
+
+	git_oid_cpy(&diff_data->old_id, &delta->old_file.id);
+	git_oid_cpy(&diff_data->new_id, &delta->new_file.id);
+
+	return 0;
+}
+
+static int binary_cb(
+	const git_diff_delta *delta,
+	const git_diff_binary *binary,
+	void *payload)
+{
+	struct diff_data *diff_data = payload;
+
+	git_buf_encode_base85(&diff_data->old_binary_base85,
+		binary->old_file.data, binary->old_file.datalen);
+	diff_data->old_binary_inflatedlen = binary->old_file.inflatedlen;
+	diff_data->old_binary_type = binary->old_file.type;
+
+	git_buf_encode_base85(&diff_data->new_binary_base85,
+		binary->new_file.data, binary->new_file.datalen);
+	diff_data->new_binary_inflatedlen = binary->new_file.inflatedlen;
+	diff_data->new_binary_type = binary->new_file.type;
+
+	return 0;
+}
+
+static int hunk_cb(
+	const git_diff_delta *delta,
+	const git_diff_hunk *hunk,
+	void *payload)
+{
+	cl_assert("did not expect hunk callback");
```

Comment:
This (and the next) assert would always succeed, since it's comparing the string against `0`. It looks like this wants to abort the test if these callbacks get called, in which case `cl_fail()` would be more appropriate.